tools/online_editor/src/editor_widget.ts

```diff
@@ -0,0 +1,551 @@
+      this._editor.layout({ width, height });
+    }
+  }
+
+  public add_model(url: string, model: monaco.editor.ITextModel) {
+    console.log("Adding model for url", url);
+    if (this._editor_documents.get(url) != null) {
+      console.log("  => Already known!");
+      return; // already know that URL
+    }
+    model.onDidChangeContent(() => {
+      this.maybe_update_preview_automatically();
+    });
+    this._editor_documents.set(url, { model, view_state: null });
+    if (this._onModelAdded != null) {
+      this._onModelAdded(url);
+    }
+    if (this._editor_documents.size === 1) {
+      this._base_url = url;
+      this.set_model(url);
+    }
+    this.update_preview();
+  }
+
+  public remove_model(url: string) {
+    console.log("Removing model for url", url);
+    this._editor_documents.delete(url);
+    if (this._onModelRemoved != null) {
+      this._onModelRemoved(url);
+    }
+  }
+
+  public set_model(url: string): boolean {
+    console.log("Setting model for url", url);
+    const model_and_state = this._editor_documents.get(url);
+    if (model_and_state != null && this._editor != null) {
+      this._editor.setModel(model_and_state.model);
+      if (model_and_state.view_state != null) {
+        this._editor.restoreViewState(model_and_state.view_state);
+      }
+      this._editor.focus();
+      if (this._onModelSelected != null) {
+        this._onModelSelected(url);
+      }
+      return true;
+    }
+    return false;
+  }
+
+  protected onResize(_msg: Message): void {
+    if (this.isAttached) {
+      this.resize_editor();
+    }
+  }
+
+  protected update_preview() {
+    const base_url = this._base_url != null ? this._base_url : "";
+    const main_model_and_state = this._editor_documents.get(base_url);
+
+    if (main_model_and_state != null) {
+      const source = main_model_and_state.model.getValue();
+
+      setTimeout(() => {
+        if (this._onRenderRequest != null) {
+          this._onRenderRequest(
+            this._style,
+            source,
+            base_url,
+            (url: string) => {
+              return this.fetch_url_content(url);
+            }
+          ).then((markers: monaco.editor.IMarkerData[]) => {
+            if (this._editor != null) {
+              const model = this._editor.getModel();
+              if (model != null) {
+                monaco.editor.setModelMarkers(model, "slint", markers);
+              }
+            }
+          });
+        }
+      }, 1);
+    }
+  }
+
+  protected maybe_update_preview_automatically() {
+    if (this.auto_compile) {
+      console.log("Refreshing auto_compile timeout.");
+      if (this._keystroke_timeout_handle != null) {
+        clearTimeout(this._keystroke_timeout_handle);
+      }
+      this._keystroke_timeout_handle = setTimeout(() => {
+        this.update_preview();
+      }, 500);
+    }
+  }
+
+  private async setup_editor(container: HTMLDivElement): Promise<void> {
+    container.className += "edit-area";
```

Comment:
I'm curious, why `+=` instead of `=` here? Perhaps you want this?
```suggestion
    container.classList.add("edit-area");
```

Reply:
Yeap! Cool that it works like this!